<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

#### Function

In [1]:
def fibonacci(n):
    if n < 1:
        return ValueError("Value must be greater than 1")
    if (n == 1):
        return 1
    elif n ==2:
        return 1 
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)

#### Simple Tests

In [2]:
fibonacci(0)

ValueError('Value must be greater than 1')

In [3]:
assert(fibonacci(1) == 1)
assert(fibonacci(2) == 1)

assert(fibonacci(3) == 2)

assert(fibonacci(5) == 5)

assert(fibonacci(8) == 21)

#### Q 1.1.2
What's the complexity of your implementation?

#### Response
The complexity of this naive implemation for generating $F_n$ is $O(2^n)$, as we are using recursion twice in each call of the function.

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

#### Function

In [4]:
def fibonacci2(n):
    if n < 1:
        return ValueError("Value must be greater than 1")
    
    storage_list = [1,1]
    
    for i in range(2, n):
        storage_list.append(storage_list[i - 1] + storage_list[i - 2])
    
    return storage_list[n - 1]

#### Simple Tests

In [5]:
fibonacci2(0)

ValueError('Value must be greater than 1')

In [6]:
assert(fibonacci2(1) == 1)
assert(fibonacci2(2) == 1)

assert(fibonacci2(3) == 2)

assert(fibonacci2(5) == 5)

assert(fibonacci2(8) == 21)

#### Q 1.1.4
What's the complexity of your implementation?

#### Response
The complexity of this implemation for generating $F_n$ is now $O(n)$, as only iterate through all values between $2$ to $n$.

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


#### Response

An example of optimization that could improve computational performance would be utilize dynamic programming and create a storage array, which is what I did in Fibonacci2, such that we do not have to recusively call functions to recalculate values that we might have already calculated. If we have storage constraits, a similar optimization method to the one I implemented in Fibonacci2 would be to only store the previous 2 values as we iterate to n, as once we've used $F_{n-2}$, we don't have to access it ever again.


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

#### Response
I appoximated $\pi$ using a Monte Carlo estimate, where I used the fact that the ratio of the area of a circle with radius 0.5 to the area of a square with side length 1 is approximately equivalent to the ratio of radomly uniformally generated finite points in the circle to the ratio of radomly uniformally generated finite points in the square. Mathematically:

$$\frac{\text{Area of Circle of radius 0.5}}{\text{Area of Square with side length 1}} \approx \frac{\text{Number of randomly generated points in Circle}}{\text{Number of randomly generated points in Square}}$$

$$\frac{\pi}{4} \approx \frac{\text{Number of randomly generated points in Circle}}{\text{Number of randomly generated points in Square}}$$

$$\pi \approx 4 * \frac{\text{Number of randomly generated points in Circle}}{\text{Number of randomly generated points in Square}}$$

I chose to center both the square and the circle at $(0.5, 0.5)$, and using Python's Uniform Random Number Generator for values in $[0,1)$, generated $(x,y)$ pairs and allocated them to both the circle and the square if the distance of $(x,y)$ to $(0.5, 0.5)$ was within the radius of the circle, otherwise the point was allocated to the square.

In [7]:
import random

def compute_pi(n_samples = 10000):
    points_in_circle = 0
    points_in_square = 0
    
    for i in range(n_samples):
        #generate samples
        x = random.random()
        y = random.random()
        
        #calulate distance from center
        d = ((x - 0.5) ** 2 + (y - 0.5) ** 2) ** 0.5
        
        if d < 0.5:
            points_in_circle += 1
            points_in_square += 1
        else:
            points_in_square += 1
    
    return 4 * points_in_circle/points_in_square

In [8]:
compute_pi(1000)

3.18

In [9]:
compute_pi(10000)

3.1376

In [10]:
compute_pi(100000)

3.1498

In [11]:
compute_pi(1000000)

3.144016

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


#### Response:

For me, the first indicator that we should research other methods is the strict requirement for k-nearest neighbors to be an approximation of the true $f(x)$. In a very practical setting, there will never be a situation where $N \rightarrow \infty$ and $k \rightarrow \infty$, and because of this, it would be very difficult for $\hat{f}(x) \rightarrow f(x)$. Aside from the fact that we would need vast amounts of data to make kNN be a universal approximator, it is very impractical to implement kNN in a production setting. Since kNN does not create or store a  model like gradient boosting or neural networks, we would have to continously call up all of the data to utilize the model every single time there is a new data point, which, over time, could become very unwieldy as data grows larger. kNN also suffers from being fairly sensitive to the number of features in the data, where calculating the distance for a new data point to every other data points to find the k-nearest neighors will also take a very long time as the number of features grow. 

As a result, it might be still more effient to research into other methods that can approximate nonlinearities, mostly due to the fact that kNN, while having the ability to become a universal approximator, requires an enormous amount of data to be feasible, thus becoming very inefficient and very difficult to use in practical applications.